From 095bc65c3e1e1c6234c030c60efa39d4af48f6c7 Mon Sep 17 00:00:00 2001
From: Kenneth Miles <tommymcguiver@gmail.com>
Date: Wed, 18 Feb 2026 17:15:14 +1000
Subject: [PATCH] block: restore non-freezing writable sysfs for
 queue/add_random

Commit 39a9f1c334f9 moved /sys/block/*/queue/add_random flag into queue_limits.features
so it can be set atomically (with queue frozen), and later it was converted
to QUEUE_LIM_RW_ENTRY.

This introduced the deadlock with device-mapper / udev-workers when
toggling /sys/block/*/queue/add_random.

Since add_random is just a hint for add_disk_randomness() and doesn't
require strict atomicity with I/O quiescing, switch back to QUEUE_RW_ENTRY.

Commit 935a20d1bebf loosened this behaviour for other sysfs attributes,
similar fix provided for this attribute.
---
 block/blk-sysfs.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/block/blk-sysfs.c b/block/blk-sysfs.c
index 003aa684e854..9766bc0e5804 100644
--- a/block/blk-sysfs.c
+++ b/block/blk-sysfs.c
@@ -349,7 +349,6 @@ static int queue_##_name##_store(struct gendisk *disk,			\
 }
 
 QUEUE_SYSFS_FEATURE(rotational, BLK_FEAT_ROTATIONAL)
-QUEUE_SYSFS_FEATURE(add_random, BLK_FEAT_ADD_RANDOM)
 QUEUE_SYSFS_FEATURE(iostats, BLK_FEAT_IO_STAT)
 QUEUE_SYSFS_FEATURE(stable_writes, BLK_FEAT_STABLE_WRITES);
 
@@ -363,6 +362,28 @@ static ssize_t queue_##_name##_show(struct gendisk *disk, char *page)	\
 QUEUE_SYSFS_FEATURE_SHOW(fua, BLK_FEAT_FUA);
 QUEUE_SYSFS_FEATURE_SHOW(dax, BLK_FEAT_DAX);
 
+
+static ssize_t queue_add_random_show(struct gendisk *disk, char *page)
+{
+       return queue_var_show(!!(disk->queue->limits.features & BLK_FEAT_ADD_RANDOM), page);
+}
+
+static ssize_t queue_add_random_store(struct gendisk *disk,
+                                     const char *page, size_t count)
+{
+       bool val;
+       int ret = kstrtobool(page, &val);
+       if (ret < 0)
+               return ret;
+
+       if (val)
+               disk->queue->limits.features |= BLK_FEAT_ADD_RANDOM;
+       else
+               disk->queue->limits.features &= ~BLK_FEAT_ADD_RANDOM;
+
+       return count;
+}
+
 static ssize_t queue_poll_show(struct gendisk *disk, char *page)
 {
 	if (queue_is_mq(disk->queue))
@@ -635,7 +656,7 @@ static struct queue_sysfs_entry queue_hw_sector_size_entry = {
 
 QUEUE_LIM_RW_ENTRY(queue_rotational, "rotational");
 QUEUE_LIM_RW_ENTRY(queue_iostats, "iostats");
-QUEUE_LIM_RW_ENTRY(queue_add_random, "add_random");
+QUEUE_RW_ENTRY(queue_add_random, "add_random");
 QUEUE_LIM_RW_ENTRY(queue_stable_writes, "stable_writes");
 
 #ifdef CONFIG_BLK_WBT
-- 
2.50.1 (Apple Git-155)

